File: julia/server/json_server.jl
#
# JSON Server with julia
#
# author: Atsushi Sakai
#

using HTTP
using JSON
using Dates

function show_http_request(http)
	@show http.message
	@show HTTP.header(http, "Content-Type")
	while !eof(http)
		println("body data: ", String(readavailable(http)))
	end
end

function get_config()
    ip = "0.0.0.0" #default
    if haskey(ENV,"SERVER_IP")
        ip = ENV["SERVER_IP"] 
    end

	port = 8000 #default
    if haskey(ENV,"SERVER_PORT")
        port =parse(Int64, ENV["SERVER_PORT"])
    end

    println("serving at ip:", ip)
    println("serving at port:", port)

    return ip, port
end


function main()
    println(PROGRAM_FILE," start!!")

    ip, port = get_config()
    
	HTTP.listen(ip, port) do http::HTTP.Stream

		show_http_request(http)

		#response json
	    jdic = Dict("title"=>"Julia JSON server", "PI" => pi)
    	jdic["time"] = Dates.format(Dates.now(), "yyyymmddHHMMSS")

		write(http, JSON.json(jdic))
	end

    println(PROGRAM_FILE," Done!!")
end

main()

File: python/server/json_server.py
"""

JSON Server with python

author: Atsushi Sakai (@Atsushi_twi)

"""

from http.server import HTTPServer, SimpleHTTPRequestHandler
from urllib.parse import parse_qs, urlparse
import json
from datetime import datetime
import math


class MyHandler(SimpleHTTPRequestHandler):

    def do_POST(self):

        # get request
        content_len = int(self.headers.get('content-length'))
        req = json.loads(self.rfile.read(content_len))
        print("request")
        print(req)

        uri = self.path
        ret = parse_qs(urlparse(uri).query, keep_blank_values=True)

        res = {}
        res["title"] = "Python JSON server"
        res["time"] = datetime.now().strftime("%Y%m%d%H%M%S")
        res["PI"] = math.pi
        print("response")
        print(res)

        ret = json.dumps(res)

        body = bytes(ret, 'utf-8')
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-length', len(body))
        self.end_headers()
        self.wfile.write(body)


def main():
    print("Start python json server")
    host = '0.0.0.0'
    port = 8000
    httpd = HTTPServer((host, port), MyHandler)
    print('serving at port', port)
    httpd.serve_forever()

    print(__file__ + " start!!")


if __name__ == '__main__':
    main()
